Adding People & SetFireType

DasKeifer opened the pull request:
Added get & set for people 1 & 2 (populated and evacuated) to allow removing/moving buildings without affecting score.
Also added SetFireType (GetFireType was already there but not set)

DasKeifer committed:
First cut that's working. Needs some cleaning

DasKeifer committed:
Cleaning up people1

DasKeifer committed:
Minor fixes and adding SetFireType fn that was missing (GetFireType is already present and setFireType was used internally)

DasKeifer committed:
Adding tests for people1

DasKeifer committed:
Adding test for fire type

KnightMiner commented:
people1 is not a good name, its really unclear what that means. A more descriptive name is preferable.

DasKeifer commented:
> people1 is not a good name, its really unclear what that means. A more descriptive name is preferable.

I agree its not a good name but its the name the game uses internally and in the save data which is why i used it. I am open to suggestions but I was thinking it would be best to make it match the game data. First name that comes to my mind that would be a good description would be BuildingScore or BuildingPeopleCount. I would lean away from using populated/population because that is already used for determining if the lights are on or not and if a destroyed building is counted against your score so I feel it could be more confusing

DasKeifer committed:
Updating exposed board functions to be cleaner and more complete

DasKeifer committed:
Fixing tests to make them more stable and passing consistantly

DasKeifer committed:
Cleaning up some more terminology and making scanning stable

DasKeifer committed:
Removing dependency on tiles being buildings for scan. Its technically not needed

Lemonymous commented:
I know I have given you some recommendations earlier, and the following might seem contradicting. I apologize for that.

Overall, I think people1 might be unnecessarily confusing, unless there is such a thing as people2, which it implies.
I see your reasoning of it being called that in the save data, but if there is no such thing as people2, then just using `people` is probably preferable.

### Board functions:
Some of the board functions that are added look slightly non-standard. Both in names and functionality. Can you provide explanations for why the set functions for example should return values, rather than be simple set functions? "Unset" is also non-standard.

Otherwise, my recommendation is to make them simpler:
```
- Board.SetScoredBuilding
- Board.UnsetScoredBuilding
- Board.GetScoredBuildingScore
+ Board.SetPeople
+ Board.GetPeople
```
My recommendation would also be to have `Board.SetFireType` not call SetTerrain or do any specific logic within it, and have it just as a pass-through to `memedit.board.setFireType`

### Tests
The test for `people` warps the testing suite around itself, and dictates that `Tests.GetCleanTile` has to change in order for `testsuite.test_tile_people1` to pass. That is the sign of a brittle test.
This test will only pass if there is at least one building tile that has not been tampered with, so `Tests.GetCleanTile` has been changed to not tamper with buildings.
If additional tests are added that do tamper with buildings, the test will fail, due to it not being built to pass from any random board state.
Changing `Tests.GetCleanTile`, and forcing the entire test suite to treat buildings carefully makes for a brittle test suite.

A better test might be to take advantage of a vanilla function that is known to change `people`. `Board.SetPopulation` is one such function. You can change the memory for people to whatever you want, then `Board.SetPopulation(true,p)` followed by `Board.SetPopulation(false,p)`, which should evacuate all the people. That would give you two distinct checks for when there are people, and when there are no people from the same `memedit.board.getPeople` function.

`test_SetFireType`
I know there are some long tests that have been written before, but I don’t think we should treat those as a templates for how to write new tests, because they are hard to read. I don't have a perfect example for you right now though, so don't worry too much about this I suppose. Ideally the tests should be easy to verify that it is written correctly, so the code clearly shows what the expected result should be when asserting.
I think maybe the test suite could potentially benefit from a renovation at some point, though I am not sure if and when I might get around to it.

#### Minor issues:
I understand that `Uncleaned` in `GetUncleanedBuilding` was referring to the opposite of `Clean` in `Tests.GetCleanTile`. The wording feels kind of awkward.
Suggestions:
```
GetUncleanedBuilding    -> GetRandomBuilding
GetUncleanedNonBuilding -> GetRandomNonBuilding
```

Variable names in `test_SetFireType` should be camelCase rather than snake_case.

DasKeifer commented:
# People 
There is a `people2` in the game and data which is why I used `people1`. From what I could tell though, it's always 0 and not used. I'm fine with calling it just `people` 

# Scans and tests
I understand the brittleness being a concern - I had issues with the chen scan too when testing. I was seeing this new scan & test fail randomly sometimes which is why I was trying to make it more stable. While I was doing that, realized it doesn't actually matter of the tile is a building or not for the scan at all. The last revision doesn't require a building and with the order it's typically run in, the board doesn't have buildings as they were all cleared by the previous tile scans. I left some of the other stuff I had added to make it more stable previously because it seemed like good things to have to keep the other tests/scans from impacting as much. With this last revision, I haven't had the new scan or test fail yet. Did you see it fail with the latest changes?

For the unique building test I modified it to clean up afterwards so it doesn't leave a building with no people as this could mess up the people test if we randomly selected the building with no people score to check the people of. To me I feel this should be kept as it is making the existing test more isolated. 

The change I made to get clean tile is because we can't fully clean a building tile since it holds onto the people value. I was thinking the idea was to get a fully clean, "blank" tile which would mean excluding buildings because of this. I was seeing the test randomly select a tile that was previously a cleared building so the test would fail since the tile still held its people value.

I'll check the `setPopulated` value but IIRC it doesn't change the people value - just whether it's counted even if destroyed or not at the end of the mission. As far as I could tell there isn't a vanilla function that affects this value. I could modify the test to set the value and read instead of starting with reading the "natural" values but that seems like less good of a test as it would not verify we are using the right value, just that we can set and and read whatever value the scan found. I guess I could also do an m of n type test but again that seems less ideal to me

# Functions 
For the functions I was trying to make them more convenient to use. I had the returns do you don't have to first get the current terrain or people value when setting/unsetting the tiles. I don't particularly care what these functions look like and will change those back. From what I could tell the people value seems to be ignored if it's not a building so it seems safe enough just to wrap the raw function and then it can use the standard functions naming.

# Final note
I'll make these changes shortly. I'll check out the `setpopulated` and get back to you on that and then we can figure out how you want to handle the tests and related changes. Probably just on discord. 

Thanks for taking the time to do this. I really appreciate it

DasKeifer commented:
SetPopulated does change the people1 value. Not sure why I missed this before. This also revealed what people2 is used for as when the building is evac'ed, it moves the value from people1 to people2.

Working to rework this PR based on this to include people2 and use setPopulated to be a better test

DasKeifer committed:
Updating to go back to simpler fns and consistent naming patterns

DasKeifer committed:
Renaming, adding people evacuated, and redoing the scan/test to use Board:SetPopulated

DasKeifer committed:
Running and fixing changes to use built in board populated functions

Lemonymous reviewed (approved):
I have reviewed the PR, and tested the following functions, which returned the expected results.
- Board:GetPeoplePopulated
- Board:SetPeoplePopulated
- Board:GetPeopleEvacuated
- Board:SetPeoplePopulated

I ran the tests in Test Mech Scenario which all passed. Some odd function sometimes failed, but that is nothing new. Unrelated to this PR.
I ran memedit calibration which also found the addresses first try.

The code looks good to me.
